File: app/src/main/java/com/tobiasfried/brewkeeper/viewmodel/EntryViewModel.java
package com.tobiasfried.brewkeeper.viewmodel;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;
import com.tobiasfried.brewkeeper.constants.IngredientType;
import com.tobiasfried.brewkeeper.model.Brew;
import com.tobiasfried.brewkeeper.model.Ingredient;

import java.util.List;

import javax.annotation.Nullable;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

public class EntryViewModel extends ViewModel {

    private static final String LOG_TAG = EntryViewModel.class.getSimpleName();

    private DocumentReference mDocRef;

    private MutableLiveData<Brew> mBrew = new MutableLiveData<>();
    private MutableLiveData<List<Ingredient>> mTeas = new MutableLiveData<>();
    private MutableLiveData<List<Ingredient>> mFlavors = new MutableLiveData<>();

    /**
     * ViewModel Constructor
     * @param database DB instance
     * @param brewId ID of edited Brew
     */
    public EntryViewModel(FirebaseFirestore database, String c, String brewId) {
        // Get Brew
        if (brewId == null) {
            mDocRef = database.collection(c).document();
            mBrew.setValue(new Brew());
        } else {
            mDocRef = database.collection(c).document(brewId);
            mDocRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                @Override
                public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                    if (task.isSuccessful() && task.getResult() != null) {
                        mBrew.setValue(task.getResult().toObject(Brew.class));
                    }
                }
            });
        }
        // Get all Teas
        database.collection(Ingredient.COLLECTION)
                .whereEqualTo("type", IngredientType.TEA)
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
            @Override
            public void onEvent(@Nullable QuerySnapshot queryDocumentSnapshots, @Nullable FirebaseFirestoreException e) {
                if (queryDocumentSnapshots != null && e == null) {
                    mTeas.setValue(queryDocumentSnapshots.toObjects(Ingredient.class));
                }
            }
        });
        // Get all Ingredients
        database.collection(Ingredient.COLLECTION)
                .whereEqualTo("type", IngredientType.FLAVOR)
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
            @Override
            public void onEvent(@Nullable QuerySnapshot queryDocumentSnapshots, @Nullable FirebaseFirestoreException e) {
                if (queryDocumentSnapshots != null && e == null) {
                   mFlavors.setValue(queryDocumentSnapshots.toObjects(Ingredient.class));
                }
            }
        });
    }

    //

    // Getters

    public DocumentReference getDocumentReference() {
        return mDocRef;
    }

    public LiveData<Brew> getBrew() {
        return mBrew;
    }

    public LiveData<List<Ingredient>> getTeas() {
        return mTeas;
    }

    public LiveData<List<Ingredient>> getFlavors() {
        return mFlavors;
    }

}

File: app/src/main/java/com/tobiasfried/rasa/entry/EntryViewModel.kt
package com.tobiasfried.rasa.entry

import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import com.tobiasfried.rasa.constants.IngredientType
import com.tobiasfried.rasa.domain.Brew
import com.tobiasfried.rasa.domain.Ingredient
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel

class EntryViewModel
/**
 * ViewModel Constructor
 * @param database DB instance
 * @param brewId ID of edited Brew
 */
(database: FirebaseFirestore, c: String, brewId: String?) : ViewModel() {

    //

    // Getters

    var documentReference: DocumentReference? = null
        private set

    private val mBrew = MutableLiveData<Brew>()
    private val mTeas = MutableLiveData<List<Ingredient>>()
    private val mFlavors = MutableLiveData<List<Ingredient>>()

    val brew: LiveData<Brew>
        get() = mBrew

    val teas: LiveData<List<Ingredient>>
        get() = mTeas

    val flavors: LiveData<List<Ingredient>>
        get() = mFlavors

    init {
        // Get Brew
        if (brewId == null) {
            documentReference = database.collection(c).document()
            mBrew.setValue(Brew())
        } else {
            documentReference = database.collection(c).document(brewId)
            documentReference!!.get().addOnCompleteListener { task ->
                if (task.isSuccessful && task.result != null) {
                    mBrew.value = task.result!!.toObject(Brew::class.java)
                }
            }
        }
        // Get all Teas
        database.collection(Ingredient.COLLECTION)
                .whereEqualTo("type", IngredientType.TEA)
                .addSnapshotListener { queryDocumentSnapshots, e ->
                    if (queryDocumentSnapshots != null && e == null) {
                        mTeas.value = queryDocumentSnapshots.toObjects(Ingredient::class.java)
                    }
                }
        // Get all Ingredients
        database.collection(Ingredient.COLLECTION)
                .whereEqualTo("type", IngredientType.FLAVOR)
                .addSnapshotListener { queryDocumentSnapshots, e ->
                    if (queryDocumentSnapshots != null && e == null) {
                        mFlavors.value = queryDocumentSnapshots.toObjects(Ingredient::class.java)
                    }
                }
    }

    companion object {

        private val LOG_TAG = EntryViewModel::class.java.simpleName
    }

}
